src/HTMLTokenizer/HTMLTokenizer.cc

```diff
@@ -0,0 +1,281 @@
+        state = this->_currentState;
+
+    Handle<Object> ret = NanNew<Object>();
+
+    switch(this->_currentState) {
+
+    case CONTENT: this->content(); break;
+    case COMPONENT: this->component(); break;
+    case COMMENT: this->comment(); break;
+
+    case END:
+    case ILLEGAL:
+        ret->Set(NanNew<String>("state"),NanNew<Number>(state));
+        ret->Set(NanNew<String>("value"),NanNull());
+        return ret;
+    break;
+
+    default:
+        this->initial();
+        return _next();
+    }
+
+    ret->Set(NanNew<String>("state"),NanNew<Number>(state));
+    ret->Set(NanNew<String>("value"),
+             NanNew<String>(this->_source.substr(start,
+                                                 this->_currentIndex - start).c_str()));
+
+    return ret;
+
+}
+void HTMLTokenizer::initial() {
+    if (this->_currentIndex >= this->_source.length()) {
+		this->_currentState = END;
+		return;
+    }
+
+    // maybe comment or component
+    if (this->_source[this->_currentIndex] == '<') {
+		// comment
+		if (this->_source[this->_currentIndex + 1] == '!') {
+			if (this->_source[this->_currentIndex + 2] == '-' &&
+				this->_source[this->_currentIndex + 3] == '-') {
+				this->_currentState = COMMENT;
+				return;
+			}
+
+			this->_currentState = CONTENT;
+			return;
+		}
+
+		if (this->checkIfComponent()) {
+			this->_currentState = COMPONENT;
+			return;
+		}
+	}
+
+    this->_currentState = CONTENT;
+
+}
+
+void HTMLTokenizer::component() {
+    this->_currentIndex += 5;
+    while (this->_currentIndex < this->_source.length()) {
+        if (this->_source[this->_currentIndex] == '>') {
+            this->_currentIndex++;
+            this->_currentState = INITIAL;
+            return;
+        }
+        this->_currentIndex++;
+    }
+
+    this->_currentState = ILLEGAL;
+
+}
+
+void HTMLTokenizer::content() {
+    this->_currentIndex++;
+	while (this->_currentIndex < this->_source.length()) {
+		if (this->_source[this->_currentIndex] == '<') {
+			this->_currentState = INITIAL;
+			return;
+		}
+		this->_currentIndex++;
+	}
+	this->_currentState = END;
+}
+
+
+
+void HTMLTokenizer::comment() {
+	this->_currentIndex += 4;
+
+	while (this->_currentIndex < this->_source.length()) {
+		if (this->_source[this->_currentIndex] == '-') {
+			if (this->_currentIndex + 2 >= this->_source.length()) {
+				this->_currentState = ILLEGAL;
+				return;
+			}
+
+			if (this->_source[this->_currentIndex + 1] == '-' &&
+				this->_source[this->_currentIndex + 2] == '>') {
+				this->_currentIndex += 3;
+				this->_currentState = INITIAL;
+				return;
+			}
+		}
+		this->_currentIndex++;
+	}
+	this->_currentState = ILLEGAL;
+
+
+}
+
+bool is_space_or_end(char c) {
+    return
+        c == ' ' ||
+        c == '\f' ||
+        c == '\n' ||
+        c == '\r' ||
+        c == '\t' ||
+        c == '\v' ||
+        c == '/' ||
+        c == '>';
+        
+}
+bool is_cat(std::string s) {
+    return
+        s[0] == '<' &&
```

Comment:
@ForNeVeR Any ideas how we can improve the performance of this code? Because according my benchmarks, this native code is slower than previous JavaScript code. It's very unexpected, but I think it's related to interop issues (V8 to C++ types conversion).